aiida/orm/node/process/calculation/calculation.py

```diff
@@ -10,4 +10,25 @@
 class CalculationNode(ProcessNode):
     """Base class for all nodes representing the execution of a calculation process."""
     # pylint: disable=too-few-public-methods,abstract-method
-    pass
+
+    _cacheable = True
+
+    def validate_incoming(self, source, link_type, link_label=None):
+        """
+        Validate adding a link of the given type from a given node to ourself.
+
+        This function will first validate the types of the inputs, followed by the node and link types and validate
+        whether in principle a link of that type between the nodes of these types is allowed.the
+
+        Subsequently, the validity of the "degree" of the proposed link is validated, which means validating the
+        number of links of the given type from the given node type is allowed.
+
+        :param source: the node from which the link is coming
+        :param link_type: the type of link
+        :param link_label: optional link label
+        :raise TypeError: if `source` is not a Node instance or `link_type` is not a `LinkType` enum
+        :raise ValueError: if the proposed link is invalid
+        """
+        super(CalculationNode, self).validate_incoming(source, link_type, link_label)
+        from aiida.orm.utils.links import validate_link
+        validate_link(source, self, link_type, link_label)
```

Comment:
Is this needed?  Not already in `Node`?

Reply:
I remember now, this needs to be in the subclasses `Data`, `CalculationNode` and `WorkflowNode`, because the tests currently, for simplicity, use base `Nodes` to create dummy graphs. If I were to put these checks in the `Node` base class, the type checks would fail. Until we rewrite the tests to only ever use `Data`, `CalculationNode` and `WorkflowNode` instances to create dummy graphs, this has to stay in the sub classes.

Reply:
Fair one